### PR TITLE
bpo-29534: _decimal difference with _pydecimal

### PR DESCRIPTION
**Pull request by andrewnester**
Fix for https://bugs.python.org/issue29534

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in "Your Details" at b.p.o
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/)
4. If you just signed the CLA, please **wait at least a day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

**Comment by mdickinson** [comment]
The new behaviour seems like the right thing to me. Can you rework the code to avoid the repetition of the `if cls is Decimal` block?

**Comment by andrewnester** [comment]
@mdickinson thanks for the comment! I've just reworked my PR to avoid code duplication.

File: Lib/_pydecimal.py [hunk 1]
```diff
@@ -734,18 +734,23 @@ def from_float(cls, f):
...
         """
         if isinstance(f, int):                # handle integer inputs
-            return cls(f)
-        if not isinstance(f, float):
-            raise TypeError("argument must be int or float.")
-        if _math.isinf(f) or _math.isnan(f):
-            return cls(repr(f))
-        if _math.copysign(1.0, f) == 1.0:
-            sign = 0
+            sign = 0 if f >=0 else 1
```
Comment (mdickinson) [inline Lib/_pydecimal.py]: Style nit: please add a space after the '>='.

Comment (mdickinson) [inline Lib/_pydecimal.py]: Style nit: please add a space after the `>=`.

File: Lib/test/test_decimal.py [hunk 2]
```diff
@@ -5467,6 +5477,7 @@ def __abs__(self):
             self.assertEqual(Decimal.from_float(cls(101.1)),
                              Decimal.from_float(101.1))
 
+
```
Comment (mdickinson) [inline Lib/test/test_decimal.py]: This change seems unrelated to the PR; please could you revert it?

File: Lib/test/test_decimal.py [hunk 3]
```diff
@@ -1185,6 +1185,16 @@ def test_wide_char_separator_decimal_point(self):
         self.assertEqual(format(Decimal('100000000.123'), 'n'),
                          '100\u066c000\u066c000\u066b123')
 
+    def test_decimal_from_float_repr(self):
```
Comment (mdickinson) [inline Lib/test/test_decimal.py]: This is just nitpicking, but I think we're testing the wrong thing here: the change is in the type of the object that's passed to the `Decimal` subclass; the `repr` of the value was just a convenient way of demonstrating the difference in behaviours between the Python and C decimal implementations. How about recording and checking the type instead of the repr?

**Review by mdickinson — COMMENTED** [review]
Changes mostly LGTM, modulo a couple of nitpicks.

Still needs:
- [ ] contributor agreement for Andrew
- [ ] Misc/NEWS entry

File: Lib/test/test_decimal.py (hunk 3 above)
Comment (andrewnester) [reply Lib/test/test_decimal.py]: @mdickinson good idea, thanks!

**Comment by andrewnester** [comment]
@mdickinson thanks! I've fixed all the nitpickings you mentioned. Also updated NEWS.

**Comment by mdickinson** [comment]
@andrewnester: Thank you! One last nitpick: please could you change the name of the test?

Apart from that, this looks good to go. Did you fill out the CLA?

**Comment by andrewnester** [comment]
@mdickinson fixed. yes, I filled it couple of hours ago :)

**Comment by mdickinson** [comment]
In, that case, LGTM unconditionally.

**Review by mdickinson — APPROVED** [review]
LGTM

**Comment by mdickinson** [comment]
The coverage gods seem a bit peeved, but I can't make head or tail of the coverage report. *Which* 6 lines are new misses? (And which 5 lines are new hits?)

**Comment by mdickinson** [comment]
Merged. Thanks for the contribution!
